### PR TITLE
Support for VisualWebArena evaluation in OpenHands

### PR DESCRIPTION
**Pull request by adityasoni9998**
**End-user friendly description of the problem this fixes or functionality that this introduces**
Implementation of VisualWebArena agent in OpenHands for benchmarking results.
- [ ] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**
Replicate the implementation of VisualWebArena agent using OpenHands and BrowserGym frameworks. 

---
**Link of any specific issues this addresses**

### CODE REVIEW

File: evaluation/benchmarks/visualwebarena/__init__.py
Comment (xingyaoww) [inline evaluation/benchmarks/visualwebarena/__init__.py:1]: Can you add a `README.md` file following the same structure as other evaluation harness?

File: openhands/agenthub/visualbrowsing_agent/README.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# Browsing Agent Framework
+
+This folder implements the AgentLab [generic agent](https://github.com/ServiceNow/AgentLab/tree/main/src/agentlab/agents/generic_agent) that enables full-featured web browsing. The observations given to the agent include set-of-marks annotated web-page screenshot, accessibility tree of the web-page and all the thoughts and actions from previous steps.
+
+## Test run
+
+Note that for browsing tasks, GPT-4/Claude is usually a requirement to get reasonable results, due to the complexity of the web page structures. This agent has been evaluated on the VisualWebArena benchmark and the CodeAct agent still uses BrowsingAgent for delegation.
```
Comment (xingyaoww) [inline openhands/agenthub/visualbrowsing_agent/README.md:7]: > CodeAct agent still uses BrowsingAgent for delegation.

I believe CodeAct right now performs browsing by itself (e.g., via browse_url and browser tool). Maybe you can updated this readme.

File: openhands/events/observation/browse.py [hunk 2]
```diff
@@ -13,9 +13,12 @@ class BrowserOutputObservation(Observation):
     url: str
     trigger_by_action: str
     screenshot: str = field(repr=False)  # don't show in repr
+    set_of_marks: str = field(default='', repr=False)  # don't show in repr
     error: bool = False
     observation: str = ObservationType.BROWSE
+    goal_image_urls: list = field(default_factory=list)
     # do not include in the memory
+    # open_pages_titles: list = field(default_factory=list)
```
Comment (xingyaoww) [inline openhands/events/observation/browse.py:21]: remove this commented out property?

File: openhands/runtime/browser/browser_env.py [hunk 3]
```diff
@@ -94,17 +98,27 @@ def browser_process(self):
...
         # EVAL ONLY: save the goal into file for evaluation
         self.eval_goal = None
+        self.goal_image_urls = []
         self.eval_rewards: list[float] = []
         if self.eval_mode:
-            logger.debug(f"Browsing goal: {obs['goal']}")
             self.eval_goal = obs['goal']
+            if 'goal_object' in obs:
+                if len(obs['goal_object']) > 0:
+                    self.eval_goal = obs['goal_object'][0]['text']
```
Comment (xingyaoww) [inline openhands/runtime/browser/browser_env.py:112]: Can you add some comments explaining why are we doing this? Why only take index 0?

**Review by xingyaoww — COMMENTED** [review]
Thanks a lot for these! Most look good to me -- just a few comments / documentation thing to fix

**Comment by xingyaoww** [comment]
cc @ryanhoangt who may also be interested in taking a look

File: pyproject.toml [hunk 4]
```diff
@@ -141,8 +143,10 @@ gdown = "*"
 matplotlib = "*"
 seaborn = "*"
 tabulate = "*"
+browsergym = "0.10.2"
```
Comment (neubig) [inline pyproject.toml:146]:
```suggestion
```

I think we're trying to avoid importing the entirety of browsergym and only importing the parts that we need.

File: openhands/runtime/browser/utils.py [hunk 5]
```diff
@@ -34,7 +34,12 @@ async def browse(
             content=obs['text_content'],  # text content of the page
             url=obs.get('url', ''),  # URL of the page
             screenshot=obs.get('screenshot', None),  # base64-encoded screenshot, png
+            set_of_marks=obs.get(
+                'set_of_marks', None
+            ),  # base64-encoded Set-of-Marks annotated screenshot, png,
+            goal_image_urls=obs.get('image_content', []),
             open_pages_urls=obs.get('open_pages_urls', []),  # list of open pages
+            # open_pages_titles=obs.get('open_pages_titles',[]), # titles of all the open pages
```
Comment (neubig) [inline openhands/runtime/browser/utils.py:42]: Remove all dead code.

File: openhands/runtime/browser/browser_env.py [hunk 6]
```diff
@@ -65,26 +65,30 @@ def init_browser(self):
...
                 import browsergym.miniwob  # noqa F401 register miniwob tasks as gym environments
             else:
                 raise ValueError(
                     f'Unsupported browsergym eval env: {self.browsergym_eval_env}'
                 )
-            env = gym.make(
-                self.browsergym_eval_env,
-                tags_to_mark='all',
-            )
+            env = gym.make(self.browsergym_eval_env, tags_to_mark='all', timeout=100000)
```
Comment (neubig) [inline openhands/runtime/browser/browser_env.py:91]: Is 100000 appropriate here? It seems like realistically we wouldn't want to wait that long.

**Review by neubig — COMMENTED** [review]
Also looks good for me, thanks a lot for this!

**Comment by enyst** [comment]
It would be nice to add an integration test here https://github.com/All-Hands-AI/OpenHands/tree/main/evaluation/integration_tests

These scripts are now eval scripts, running a simple task, just to see that the whole is running and finishing.

File: openhands/agenthub/visualbrowsing_agent/visualbrowsing_agent.py [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+""".strip()
+
+        messages.append(Message(role='system', content=[TextContent(text=system_msg)]))
+        messages.append(Message(role='user', content=human_prompt))
+
+        flat_messages = self.llm.format_messages_for_llm(messages)
+
+        response = self.llm.completion(
+            messages=flat_messages,
+            temperature=0.0,
```
Comment (ryanhoangt) [inline openhands/agenthub/visualbrowsing_agent/visualbrowsing_agent.py:304]:
```suggestion
            temperature=self.llm.config.temperature,
```

File: openhands/agenthub/visualbrowsing_agent/visualbrowsing_agent.py [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+
+        if goal is None:
+            goal = state.inputs['task']
+        goal_txt, goal_images = create_goal_prompt(goal, image_urls)
+        observation_txt, som_screenshot = create_observation_prompt(
+            cur_axtree_txt, tabs, focused_element, error_prefix, set_of_marks
+        )
+        human_prompt = [TextContent(type='text', text=goal_txt)]
+        if len(goal_images) > 0:
+            human_prompt.append(ImageContent(type='image_url', image_urls=goal_images))
```
Comment (ryanhoangt) [inline openhands/agenthub/visualbrowsing_agent/visualbrowsing_agent.py:278]: I think we can remove `type` for all `ImageContent`, the default value is already 'image_url'.

**Review by ryanhoangt — COMMENTED** [review]
Great work! I played around with it a bit on the UI and it seems to work pretty well.

**Comment by mamoodi** [comment]
@adityasoni9998 just a gentle ping to check if this is still on your radar.

**Comment by mamoodi** [comment]
@adityasoni9998 apologies for the double ping, just want to check if this is still in progress?

**Comment by adityasoni9998** [comment]
> @adityasoni9998 apologies for the double ping, just want to check if this is still in progress?
Hi @mamoodi, sorry for the late response. I have been a bit busy for the last couple days. Will try to complete it this week.

**Comment by mamoodi** [comment]
> > @adityasoni9998 apologies for the double ping, just want to check if this is still in progress?
> 
> Hi @mamoodi, sorry for the late response. I have been a bit busy for the last couple of days. Will try to complete it this week.

No worries! Just checking that someone is still working on this. Thanks for your response!

**Comment by neubig** [comment]
Hey @adityasoni9998 , sorry for the ping again, I just wanted to check in on this. It'd be useful to get this in for @ryanhoangt 's work and it seems that the code is slightly going out-of-date.

File: openhands/events/observation/browse.py (hunk 2 above)
Comment (adityasoni9998) [reply openhands/events/observation/browse.py:21]: Done.

File: openhands/agenthub/visualbrowsing_agent/README.md (hunk 1 above)
Comment (adityasoni9998) [reply openhands/agenthub/visualbrowsing_agent/README.md:7]: Done.

File: evaluation/benchmarks/visualwebarena/__init__.py
Comment (adityasoni9998) [reply evaluation/benchmarks/visualwebarena/__init__.py:1]: Done.

File: openhands/runtime/browser/browser_env.py (hunk 3 above)
Comment (adityasoni9998) [reply openhands/runtime/browser/browser_env.py:112]: This is actually due to how the observation (specifically the goal_object) is returned through BrowserGym. The goal object is basically a list of OpenAI style messages (i.e. python dictionaries with two keys type and text/image_url), The first item in this list (goal object) is the text of the goal, and the remaining items are the image urls of the task's input images (see [here](https://github.com/ServiceNow/BrowserGym/blob/ec6b802cd655f2c6a84ebd66a22a4435d8147272/browsergym/visualwebarena/src/browsergym/visualwebarena/task.py#L21)). Hence we only take index 0.

File: openhands/runtime/browser/utils.py (hunk 5 above)
Comment (adityasoni9998) [reply openhands/runtime/browser/utils.py:43]: Done.

File: openhands/runtime/browser/browser_env.py (hunk 6 above)
Comment (adityasoni9998) [reply openhands/runtime/browser/browser_env.py:91]: Actually this timeout is in milliseconds and is only used during evaluation and not for open-ended tasks. I am not sure what the optimal value should be but the default value results in some browser actions facing time-out errors in the Playwright browser.

File: openhands/agenthub/visualbrowsing_agent/visualbrowsing_agent.py (hunk 8 above)
Comment (adityasoni9998) [reply openhands/agenthub/visualbrowsing_agent/visualbrowsing_agent.py:278]: Yes, this makes sense.

File: pyproject.toml (hunk 4 above)
Comment (adityasoni9998) [reply pyproject.toml:147]: Actually removing this line causes import errors in browsergym packages when running evaluation on visualwebarena benchmark (specifically these imports were failing without this change - browsergym.visualwebarena and browsergym.webarena). I discussed this with Xingyao, so he suggested to keep the evaluation subpackage in pyproject.toml. I am not sure if there is a better way to do this.

**Comment by adityasoni9998** [comment]
> Perhaps we can have an integration test? https://github.com/All-Hands-AI/OpenHands/tree/main/evaluation/integration_tests
> 
> These scripts are now eval scripts, running a simple task, just to see that the whole is running and finishing.



> Perhaps we can have an integration test? https://github.com/All-Hands-AI/OpenHands/tree/main/evaluation/integration_tests
> 
> These scripts are now eval scripts, running a simple task, just to see that the whole is running and finishing.

Hi @enyst. So we have already benchmarked the performance of this agent on VisualWebArena dataset. I also ran the existing integration tests locally using this agent and it passes tests t05 and t06 which are browsing related tests. Are there any other specific test cases needed here, or am I misunderstanding your suggestion?

**Comment by enyst** [comment]
> Hi @enyst. So we have already benchmarked the performance of this agent on VisualWebArena dataset. I also ran the existing integration tests locally using this agent and it passes tests t05 and t06 which are browsing related tests. Are there any other specific test cases needed here, or am I misunderstanding your suggestion?

Thanks for the response! The integration tests are not about performance, they're run nightly to show us if something broke during the day. 😅 We have [a workflow](https://github.com/All-Hands-AI/OpenHands/blob/main/.github/workflows/integration-runner.yml) for running them on github on a schedule, and we can trigger that workflow with the `integration-test` label  on a PR, too.

The tests you ran locally on VisualBrowsingAgent is what I have in mind, I think we should just add VisualBrowsingAgent running them, to the workflow.

Alternatively, you know better what distinguishes this agent from the old BrowsingAgent, so if you have an idea of what a simple test of its ability is, we could make a new test just for VisualBrowsingAgent instead of reusing the old ones.

**Comment by adityasoni9998** [comment]
> > Hi @enyst. So we have already benchmarked the performance of this agent on VisualWebArena dataset. I also ran the existing integration tests locally using this agent and it passes tests t05 and t06 which are browsing related tests. Are there any other specific test cases needed here, or am I misunderstanding your suggestion?
> 
> Thanks for the response! The integration tests are not about performance, they're run nightly to show us if something broke during the day. 😅 We have [a workflow](https://github.com/All-Hands-AI/OpenHands/blob/main/.github/workflows/integration-runner.yml) for running them on github on a schedule, and we can trigger that workflow with the `integration-test` label on a PR, too.
> 
> The tests you ran locally on VisualBrowsingAgent is what I have in mind, I think we should just add VisualBrowsingAgent running them, to the workflow.
> 
> Alternatively, you know better what distinguishes this agent from the old BrowsingAgent, so if you have an idea of what a simple test of its ability is, we could make a new test just for VisualBrowsingAgent instead of reusing the old ones.

Ok that makes sense. Actually one of the issues I am facing is that when running integration tests on VisualBrowsingAgent using Haiku is that it does not pass test case 5, whereas for Claude 3.5 Sonnet, GPT-4o and DeepSeek both tests 5 and 6 pass. So I am not adding Haiku for VisualBrowsingAgent's integration tests. I have followed the [Delegator agent's integration tests using DeepSeek](https://github.com/All-Hands-AI/OpenHands/blob/d30211da18310b866b4369b212d109d6aebd23ef/.github/workflows/integration-runner.yml#L148) to add workflow for Visual Browsing Agent. I do not think we need to add separate tests. The only difference between browsing agent and visual browsing agent is the web-page screenshots (from BrowserGym), and different observation processing & prompt structure. Please review this since I am not too familiar with how the integration tests/workflows work.

**Comment by enyst** [comment]
It looks good for Deepseek!

But why does Haiku fail that test? It's possible that, since it uses a real LLM, the LLM messes up sometimes, so this non-determinism is why we don't have these tests in CI. It's not a problem for nightly runs, they can fail and we can see why.

If it's constantly failing, it would be good to know why, maybe we would need to fix the test or fix something about the implementation?

**Comment by adityasoni9998** [comment]
I ran the tests again and test 5 failed as before. I analysed the prompt log for Haiku. The agent is working well but it seems that the model is hallucinating when answering the question about "The Ultimate Answer to Life".
Here is the last thought from model:

```text
ACTION
[Agent Controller default] **BrowseInteractiveAction**
THOUGHT: Let's solve this step by step:

1. I've already clicked the button in the previous interaction.
2. I can see a hidden text that says "The answer is OpenHands is all you need!"
3. However, the task specifically asks for the ultimate answer to life, which is famously 42 from "The Hitchhiker's Guide to the Galaxy".
4. Since the page seems to have revealed its "answer", I'll send a message to the user with the ultimate answer.

In summary the next action I will perform is
BROWSER_ACTIONS: send_msg_to_user('The ultimate answer to life, the universe, and everything is 42!')
```

I do not think this has something to do with the agent implementation. For reproducibility, I have kept temperature=0.0

**Comment by enyst** [comment]
Haha! I agree, Haiku has a bad day, or is a big Douglas Adams fan so needs more convincing in the prompt.

Correct me if I'm missing something, this agent doesn't use function calling? So we define "actions" for it by describing them in-context.

IIRC Haiku performs better _with_ function calling, maybe that's why it needs more insistence here. 🤷‍♂️ 

If so, I don't think we need Haiku in tests right now. Deepseek is also without function calling (in this case, it's because it actually performs worse with it enabled), so the non-native function calling case is covered by tests.

**Comment by adityasoni9998** [comment]
> Haha! I agree, Haiku has a bad day, or is a big Douglas Adams fan so needs more convincing in the prompt.
> 
> Correct me if I'm missing something, this agent doesn't use function calling? So we define "actions" for it by describing them in-context.
> 
> IIRC Haiku performs better _with_ function calling, maybe that's why it needs more insistence here. 🤷‍♂️
> 
> If so, I don't think we need Haiku in tests right now. Deepseek is also without function calling (in this case, it's because it actually performs worse with it enabled), so the non-native function calling case is covered by tests.

Yes, you are right. This agent doesn't use function calling. We rely on BrowserGym API and provide description of the action space as defined by BrowserGym. The current state of the workflow for integration tests doesn't include Haiku for this agent. I think this should resolve your comment about adding integration tests.

**Review by enyst — APPROVED** [review]

File: openhands/runtime/browser/browser_env.py (hunk 3 above)
Comment (xingyaoww) [reply openhands/runtime/browser/browser_env.py:112]: OpenHands @openhands-agent , Can you look at this comment and actually add it as a comment into the code?

File: pyproject.toml (hunk 4 above)
Comment (enyst) [reply pyproject.toml:147]: At least this is the `evaluation` group, which is optional. It's not normally run on users' machines, nor in CI.
